File: 15. 3Sum/Fixed/3sum.kt
var result_number: Int = 0;
val number_of_elements: Int = 3;

data class Result ( var values: Array<Int>, var number_of_values: Int, var not_null: Boolean){
	/*
	var values: Array<Int>
	var number_of_values: Int
	var not_null: Boolean
	
	
	constructor (list: Array<Int>, nov: Int, nn: Boolean){
		this.values = list;
		this.number_of_values = nov;
		this.not_null = true;
	}
	*/
	
	constructor (list: Array<Int>): this(list, list.size, true){
		
	}
	
	constructor (): this(Array(number_of_elements) {0}, number_of_elements, false){
		//this(Array(number_of_elements) {0}, number_of_elements, false);
		/*
		values = Array(number_of_elements){0};
		number_of_values = number_of_elements;
		not_null = false;
		*/
	}
	
	fun set_result(list: Array<Int>){
		
		for (i in 0 until list.size){
			values[i] = list[i];
		}
	}
	
	fun print_result(){
		println("This is the handy work of Kotlin");
		for (i in 0 until number_of_values){
			print(values[i]);
			if (i < number_of_values - 1){
				print("\t");
			}
		}
		println();
	}
}

fun factorial(number: Int): Int{
	var out: Int = 1;
	
	for (i in number downTo 1){
		out *= i;
	}
	
	return (out);
}

fun sum_of_three (list: Array<Int>, target: Int, sample_space: Int): Array<Result> {
	var result_pointer: Int = 0;
	var len_list: Int = list.size;
	
	//val o: Result = Result;
	
	//var out: Array<Array<Int>> = Array(20){ Array(3) {0}};
	var out: Array<Result> = Array(sample_space) {Result()};
	
	for (a in 0 until len_list){
		for (b in 0 until len_list){
			if (b > a){
				for (c in 0 until len_list){
					if (c > b){
						if ((a != b) && (a != c) && (b != c)){
							var aa = list[a];
							var bb = list[b];
							var cc = list[c];
							
							if ((aa != bb) && (aa != cc) && (bb != cc)){
								if ((aa + bb + cc) == target){
									var q: Array<Int> = arrayOf(aa, bb, cc);
									out[result_pointer] = Result(q);
									result_pointer += 1;
									
									result_number = result_pointer;
								}
							}
						}
					}
				}
			}
		}
	}
	
	return out;
}

fun main (){
	var nums: Array<Int> = arrayOf(-1,0,1,2,-1,-4);
	
	val len_nums: Int = nums.size;
	
	val sample_space: Int = (factorial(len_nums)/(factorial(len_nums - number_of_elements) * factorial(number_of_elements)));
	
	println(sample_space);
	
	var h: Array<Result> = sum_of_three(nums, 0, sample_space);
	
	for (i in 0 until result_number){
		h[i].print_result();
	}
}

File: 15. 3Sum/Fixed/3sum.cs
using System;

namespace sum_of_three_ns
{
	class sum_of_three_class
	{
		public static int result_number = 0;
		public static int number_of_elements = 3;
		
		public class result {
			int [] values = new int [number_of_elements];
			public bool isNull = false;
			
			public result (){
				isNull = true;
				/*for (int i = 0; i < a.Length; i++){
					values[i] = a[i];
				}*/
			}
			
			public void set_result (int [] a){
				for (int i = 0; i < a.Length; i++){
					values[i] = a[i];
				}
			}
			
			public void print(){
				for (int i = 0; i < values.Length; i++){
					Console.Write(values[i]);
					Console.Write("\t");
				}
				Console.WriteLine();
			}
		}
		
		static int factorial(int number){
			int nje = 1;
			
			for (int i = number; i > 0; i--){
				nje *= i;
			}
			
			return (nje);
		}
		
		static result [] sum_of_three (int [] list, int combination, int target = 0)
		{
			int len_list = list.Length;
			result [] nje = new result [combination];
			int result_pointer = 0;
			
			for (int a = 0; a < len_list; a++){
				for (int b = 0; b < len_list; b++){
					if (b > a){
						for (int c = 0; c < len_list; c++){
							if (c > b){
								if ((a != b) && (a != c) && (b != c)){
									int aa = list[a];
									int bb = list[b];
									int cc = list[c];
									
									if ((aa != bb) && (aa != cc) && (bb != cc)){
										if ((aa + bb + cc) == target){
											//int [] z = {a, b, c};
											//int [] z = {aa, bb, cc};
											
											int [] y = {aa, bb, cc};
											result z = new result();
											z.set_result(y);
											nje [result_pointer] = z;
											result_pointer++;
											result_number = result_pointer;
											
											/*
											for (int i = 0; i < z.Length; i++){
												nje[i] = z[i];
												Console.Write(nje[i]);
												Console.Write("\t");
											}
											Console.Write("\n\n");
											*/
										}
									}
								}
							}
						}
					}
				}
			}
			return nje;
		}
		
		static void Main (string [] args)
		{
			int [] nums = {-1,0,1,2,-1,-4};
			
			int len_nums = nums.Length;
			
			int sample_space = (factorial(len_nums) / (factorial(len_nums - number_of_elements) * factorial(number_of_elements)));
			
			result [] h = sum_of_three(nums, sample_space);
			
			for (int i = 0; i < result_number; i++){
				//Console.WriteLine(h[i]);
				h[i].print();
			}
		}
		
	}
}
